### PR TITLE
Send heartbeats in StateClosing (JS)

### LINKED ISSUE #2868: Send heartbeats when connection state is StateClosing
We currently send heartbeats only when the connection state is StateActive or StateHolding:

https://github.com/zeroc-ice/ice/blob/9ab4adbe63adacacc0ed523e7ba0e6c2108cbec4/csharp/src/Ice/ConnectionI.cs#L1499

We should also send them in StateClosing.

Scenario:
  - in server, communicator shutdown transitions the state of an incoming connection to StateClosing:
 https://github.com/zeroc-ice/ice/blob/9ab4adbe63adacacc0ed523e7ba0e6c2108cbec4/csharp/src/Ice/ConnectionI.cs#L150

  - connection does not transition to StateClosingPending because there are outstanding upcalls (e.g. dispatches):
 https://github.com/zeroc-ice/ice/blob/9ab4adbe63adacacc0ed523e7ba0e6c2108cbec4/csharp/src/Ice/ConnectionI.cs#L1786

  - new client requests get discarded, with no responses
 https://github.com/zeroc-ice/ice/blob/9ab4adbe63adacacc0ed523e7ba0e6c2108cbec4/csharp/src/Ice/ConnectionI.cs#L2322

The logic here is once upcalls in the server connection reach 0, we call initiateShutdown which sends a CloseConnection frame and transitions to StateClosingPending when done:
https://github.com/zeroc-ice/ice/blob/9ab4adbe63adacacc0ed523e7ba0e6c2108cbec4/csharp/src/Ice/ConnectionI.cs#L1826
or
https://github.com/zeroc-ice/ice/blob/9ab4adbe63adacacc0ed523e7ba0e6c2108cbec4/csharp/src/Ice/ConnectionI.cs#L2109

However, this can take a while, and during this time, the server connection should keep sending heartbeats to the client connection.

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR fixes #2868 in JS.

It also removes ConnectionI.hold that was not called anywhere.

### CODE REVIEW

File: js/src/Ice/ConnectionI.js [hunk 1]
```diff
@@ -42,10 +42,13 @@ import "./CurrentExtensions.js";
 const StateNotInitialized = 0;
 const StateNotValidated = 1;
 const StateActive = 2;
+// StateHolding is a very transient state in JS: the connection is in StateHolding once the connection establishment
+// has succeeded and until the outgoing connection factory activates the connection - which is essentially immediate.
```
Comment (bernardnormier) [inline js/src/Ice/ConnectionI.js:46]: The hold -> active logic is still used by the OutgoingConnectionFactory, like in other language mappings.

This occurs only once.

File: js/src/Ice/ConnectionI.js [hunk 2]
```diff
@@ -935,42 +926,24 @@ export class ConnectionI {
...
 
                 case StateActive: {
-                    //
-                    // Can only switch from holding or not validated to
-                    // active.
-                    //
-                    if (this._state !== StateHolding && this._state !== StateNotValidated) {
-                        return;
-                    }
+                    Debug.assert(this._state <= StateHolding);
```
Comment (bernardnormier) [inline js/src/Ice/ConnectionI.js:929]: It looks like we can transition directly from StateNotValidated to StateActive, without an intermediate StateHolding.

File: js/src/Ice/ConnectionI.js [hunk 3]
```diff
@@ -1065,14 +1041,14 @@ export class ConnectionI {
     }
 
     sendHeartbeat() {
-        if (this._state == StateActive || this._state == StateHolding) {
+        if (this._state === StateActive || this._state === StateHolding || this._state === StateClosing) {
```
Comment (bernardnormier) [inline js/src/Ice/ConnectionI.js:1044]: The "core" of this PR: keep sending heartbeats in StateClosing, but stop when we reach StatePendingClosing (= after we've sent the CloseConnection message).

File: js/src/Ice/ConnectionI.js [hunk 4]
```diff
@@ -1338,12 +1319,13 @@ export class ConnectionI {
             switch (messageType) {
                 case Protocol.closeConnectionMsg: {
                     TraceUtil.traceRecv(info.stream, this._logger, this._traceLevels);
+                    // We transition directly to StateClosed, not StateClosingPending.
```
Comment (bernardnormier) [inline js/src/Ice/ConnectionI.js:1322]: In other language mappings, we transition to StateClosingHolding then (usually) immediately to StateClosed. See for example:

https://github.com/zeroc-ice/ice/blob/3fd9cdc81a2d3adbd666fb941777050cf1c8a102/csharp/src/Ice/ConnectionI.cs#L2301

There is no Transceiver.closing in JS. When comparing with other languages, look at the behavior of the TcpTransceiver.closing:

https://github.com/zeroc-ice/ice/blob/3fd9cdc81a2d3adbd666fb941777050cf1c8a102/csharp/src/Ice/Internal/TcpTransceiver.cs#L25

Note that we stop sending heartbeats once we receive CloseConnection from the peer, and StateClosingPending and StateClosed are equivalent in this respect.

File: js/src/Ice/ConnectionI.js [hunk 5]
```diff
@@ -1042,11 +1015,14 @@ export class ConnectionI {
         os.writeInt(Protocol.headerSize); // Message size.
 
         scheduleCloseTimeout(this);
-        this.sendMessage(OutgoingMessage.createForStream(os));
+
+        if ((this.sendMessage(OutgoingMessage.createForStream(os)) & AsyncStatus.Sent) > 0) {
+            this.setState(StateClosingPending);
```
Comment (bernardnormier) [inline js/src/Ice/ConnectionI.js:1020]: After that, we wait for the peer to abort the transport connection, which transitions our connection to StateClosed.

File: js/src/Ice/ConnectionI.js [hunk 6]
```diff
@@ -561,17 +556,13 @@ export class ConnectionI {
                     return;
                 }
 
-                this._transceiver.unregister();
```
Comment (pepone) [inline js/src/Ice/ConnectionI.js:564]: I guess this was redundant, setting the holding state does the same.

**Review by pepone — APPROVED** [review]

**Review by externl — APPROVED** [review]

File: js/src/Ice/ConnectionI.js (hunk 6 above)
Comment (bernardnormier) [reply js/src/Ice/ConnectionI.js:564]: Yes, and it's even more obvious now that setState(StateHolding) is called only once.

**Pull request merged**